client/rest/src/main/java/org/elasticsearch/client/ResponseListener.java

```diff
@@ -19,24 +19,22 @@
 
 package org.elasticsearch.client;
 
-import org.apache.http.StatusLine;
-import org.apache.http.client.methods.CloseableHttpResponse;
-import org.apache.http.message.BasicHttpResponse;
-
-import java.io.IOException;
-
 /**
- * Simple {@link CloseableHttpResponse} impl needed to easily create http responses that are closeable given that
- * org.apache.http.impl.execchain.HttpResponseProxy is not public.
+ * Listener to be provided when calling async performRequest methods provided by {@link RestClient}.
+ * Those methods that do accept a listener will return immediately, execute asynchronously, and notify
+ * the listener whenever the request yielded a response, or failed with an exception.
  */
-class CloseableBasicHttpResponse extends BasicHttpResponse implements CloseableHttpResponse {
+public interface ResponseListener {
 
-    public CloseableBasicHttpResponse(StatusLine statusline) {
-        super(statusline);
-    }
+    /**
+     * Method invoked if the request yielded a successful response
+     */
+    void onSuccess(Response response);
```

Comment:
> FailureListener is for getting notified whenever any request attempt fails, regardless of whether the request will finally succeed on another node. 

I don't know the details here, but this seems incorrect?  As a user, I care about whether the request succeeds or fails, not whether the internals needed to retry.  What am I supposed to do with failure responses when I'm not sure whether I'll get a success response later on?